[AutoDiff] TF-216: Fix adjoint buffer allocation logic.

dan-zheng opened the pull request:
Previously, `AdjointEmitter::getAdjointBuffer` allocated adjoint buffers at the current
SILBuilder position. Those adjoint buffers were freed at the end of the adjoint function.

However, this breaks modular allocation/deallocation in AdjointGen for instructions like
`alloc_stack` and `dealloc_stack`.

Now, adjoint local buffer allocation is modular:
- Local buffers are allocated at the beginning of the adjoint entry BB
  (done in `AdjointEmitter::getAdjointBuffer`).
- Those buffers are deallocated at the end of the adjoint exit BB
  (done in `AdjointEmitter::run`).

Remove old code for setting original indirect parameter adjoint buffers.
Minor gardening.

Resolves [TF-216](https://bugs.swift.org/browse/TF-216).

dan-zheng committed:
[AutoDiff] TF-216: Fix adjoint buffer allocation logic.

Previously, `AdjointEmitter::getAdjointBuffer` allocated adjoint buffers at the current
SILBuilder position. Those adjoint buffers were freed at the end of the adjoint function.

However, this breaks modular allocation/deallocation in AdjointGen for instructions like
`alloc_stack` and `dealloc_stack`.

Now, adjoint local buffer allocation is modular:
- Local buffers are allocated at the beginning of the adjoint entry BB
  (done in `AdjointEmitter::getAdjointBuffer`).
- Those buffers are deallocated at the end of the adjoint exit BB
  (done in `AdjointEmitter::run`).

Remove old code for setting original indirect parameter adjoint buffers.
Minor gardening.

Resolves TF-216.

dan-zheng commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -3761,28 +3750,16 @@ class AdjointEmitter final : public SILInstructionVisitor<AdjointEmitter> {
-        cleanup->applyRecursively(builder, adjLoc);
-      builder.createDeallocStack(adjLoc, adjBuf);
+    // Deallocate local allocations.
+    for (auto *alloc : localAllocations) {
```
`localAllocations` are no longer deallocated in reverse order because `localAllocBuilder` creates new adjoint buffers *before* the last allocated adjoint buffer. This is due to a quirk of `SILBuilder`: insertion point can only be set such that new instructions come before an instruction, not after.

This is all fine is `getAdjointBuffer` is the only code appending to `localAllocations`, but there is one other raw call to `localAllocations.push_back(adjBuf);` for the seed adjoint. But actually, I think that raw call should be removed - let me try to remove it now.

dan-zheng replied:
The only raw call to `localAllocations.push_back` is removed in https://github.com/apple/swift/pull/22645/commits/e90c5b7288ec26a438c59ff9c62ee47d6b9c8cec.

Now, it's fine that `localAllocations` are deallocated in forward order (because there's only one place where `localAllocations` is appended and we can reason about the ordering).

dan-zheng committed:
Clean up adjoint assignment for original result in `AdjointEmitter::run`.

rxwei commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -3761,28 +3729,16 @@ class AdjointEmitter final : public SILInstructionVisitor<AdjointEmitter> {
+      // Assert that local allocations have at least one use.
+      // Buffers should not be allocated needlessly.
+      assert(!alloc->use_empty());
+      builder.createDeallocStack(adjLoc, alloc);
```
You need to apply cleanups.

dan-zheng replied:
Done in https://github.com/apple/swift/pull/22645/commits/897fa8f73af077cecbfffe7f8f21b9a2ef0ec084.

rxwei reviewed (commented):
LGTM. Thanks.

dan-zheng committed:
Clean up local allocations.

Address feedback from @rxwei.

rxwei commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -3519,17 +3523,38 @@ class AdjointEmitter final : public SILInstructionVisitor<AdjointEmitter> {
+        b.createDestroyAddr(loc, v);
+    };
+    auto bufWithCleanup =
+        ValueWithCleanup(newBuf, makeCleanup(newBuf, cleanupFn));
```
```suggestion
    ValueWithCleanup bufWithCleanup(newBuf, makeCleanup(newBuf, cleanupFn));
```

dan-zheng commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -3530,7 +3530,8 @@ class AdjointEmitter final : public SILInstructionVisitor<AdjointEmitter> {
           getAdjoint().getEntryBlock()->begin());
     else
-      localAllocBuilder.setInsertionPoint(localAllocations.back());
+      localAllocBuilder.setInsertionPoint(
```
It would be nice not to set the insertion point in this `else` branch every time, but unfortunately it's necessary to avoid `stack dealloc does not match most recent stack alloc` verification failure.

I believe this is because the insertion point of `localAllocBuilder` gets moved back as `builder` creates instructions, so the insertion point is no longer at the top of the adjoint entry BB.

dan-zheng committed:
Minor style fix.

Address feedback by @rxwei.

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng committed:
Fix `localAllocBuilder` insertion point logic.

Handle case where adjoint entry BB is empty, to prevent crash.
Fix doc comment: `end of the adjoint entry BB` -> `start ...`

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow linux

dan-zheng commented:
@swift-ci Please test tensorflow linux

dan-zheng commented:
macOS tests flaked again.

I noticed a pattern: macOS tests fail if CI is triggered on a different PR. CI actually starts for that PR and CI for this PR fails due to "missing file" errors, ostensibly because files are overwritten.

### Evidence

* Two CI jobs running at once on macOS CI machine:

![image](https://user-images.githubusercontent.com/5590046/52851255-7ec6de00-30ca-11e9-847a-c866a163909e.png)
(image from https://ci-external.swift.org/job/swift-PR-TensorFlow-macOS)

* Example error (from [this job](https://ci-external.swift.org/job/swift-PR-TensorFlow-macOS/979/console)):
```
Testing: 0 .Exception in thread Thread-3:
Traceback (most recent call last):
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/threading.py", line 810, in __bootstrap_inner
    self.run()
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/threading.py", line 763, in run
    self.__target(*self.__args, **self.__kwargs)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/multiprocessing/pool.py", line 405, in _handle_results
    cache[job]._set(i, obj)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/lib/python2.7/multiprocessing/pool.py", line 572, in _set
    self._callback(self._value)
  File "/Volumes/BuildData/Users/swiftci/jenkins/workspace/swift-PR-TensorFlow-macOS/llvm/utils/lit/lit/run.py", line 179, in consume_test_result
  File "/Volumes/BuildData/Users/swiftci/jenkins/workspace/swift-PR-TensorFlow-macOS/llvm/utils/lit/lit/main.py", line 47, in update
  File "/Volumes/BuildData/Users/swiftci/jenkins/workspace/swift-PR-TensorFlow-macOS/llvm/utils/lit/lit/main.py", line 162, in update_incremental_cache
OSError: [Errno 2] No such file or directory: '/Volumes/BuildData/Users/swiftci/jenkins/workspace/swift-PR-TensorFlow-macOS/swift/validation-test/compiler_crashers_fixed/28248-swift-dynamicselftype-get.swift'
```

---

@shahmishal: Could you please verify whether my observations are correct? Is it possible to fix this macOS CI behavior, assuming it's not intentional? I don't believe Linux CI has this problem.

dan-zheng commented:
Merging despite flaky macOS CI (tested locally on macOS).

shahmishal commented:
@dan-zheng Fixed the two node issue.

dan-zheng commented:
> @dan-zheng Fixed the two node issue.

Awesome, thank you so much!
